File: problems/basic/49/solution.cpp
// Accepted: 30ms (65.91%), 22.80 (17.49%)

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        vector<string> sortedStrs;
        for (string s : strs) {
            sort(s.begin(), s.end());
            sortedStrs.push_back(s);
        }

        unordered_map<string, vector<string>> anagramMap;
        for (int i = 0; i < strs.size(); ++i) {
            anagramMap[sortedStrs[i]].push_back(strs[i]);
        }

        vector<vector<string>> anagramGroups;
        for (auto [str, group] : anagramMap) {
            anagramGroups.push_back(group);
        }

        return anagramGroups;
    }
};

File: problems/basic/49/solution.py
# Accepted: 91ms (83.38%), 20.78MB (42.76%)

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        sortedStrs = list(map(lambda s: "".join(sorted(s)), strs))

        anagramMap: Dict[str, List[str]] = defaultdict(list)
        for sortedString, string in zip(sortedStrs, strs):
            anagramMap[sortedString].append(string)

        anagramGroups = list(anagramMap.values())
        return anagramGroups
